### PR TITLE
Android NDK r17 support

### PR DESCRIPTION
**Pull request by jin**
[Full NDK changelog](https://github.com/android-ndk/ndk/wiki/Changelog-r17)

Changes related to Bazel's NDK toolchain/CROSSTOOL generation:

- Removed support for targeting mips, mips64, armeabi.
- Bumped bundled clang version to 6.0.2.
- The default STL is now libc++. If Bazel detects that r17 is being used, it
  aliases `@androidndk//:default_toolchain` to `@androidndk//:toolchain-libcpp`,
  instead of `@androidndk//:toolchain-gnu-libstdcpp`.
- Removed tests that build against armeabi, mips, mips64.
- Stop passing `-fno-integrated-as` into the compiler by default.

TODO:

- [x] Java tests
- [x] ~Support for wrap.sh to use with ASANs~ Implementing in follow up PR.
- [ ] Fix or disable remote caching with android_ndk_integration_test https://github.com/bazelbuild/bazel/issues/4663

Fixes https://github.com/bazelbuild/bazel/issues/4742

RELNOTES: Added support for Android NDK r17. The default STL is now `libc++`, and support for targeting `mips`, `mips64` and `ARMv5` (`armeabi`) has been removed.

### CODE REVIEW

File: src/test/shell/bazel/android/android_ndk_integration_test.sh [hunk 1]
```diff
@@ -240,12 +237,11 @@ EOF
 #include <arm_neon.h>
 int main() { return 0; }
 EOF
-  bazel build //:foo \
-    --compiler=clang5.0.300080 \
+  assert_build :foo \
+    --compiler=clang6.0.2 \
```
Comment (jin) [inline src/test/shell/bazel/android/android_ndk_integration_test.sh]: This will "pass" on CI because the test is disabled. The CI has r15, not r7.

**Comment by jin** [comment]
@ahumesky ping, PTAL at the core changes. I don't think I will add the support for wrap.sh in this PR, will follow up with that in a following CL. The diff is already quite big with 530 lines added.

**Comment by aselle** [comment]
I was wondering if there was a way that the (yearly) changes needed to make bazel support the newest version could be streamlined. It would be really cool if this could be synchronized with the android team so that the moment when android releases the ndk, bazel supports it. One of the best features of bazel is that it makes android builds easy, but once a year for about a month it isn't.

**Comment by jin** [comment]
@aselle there a few ways to make this more streamlined, a couple of which are under the Bazel's team purview:

1) migrate android_ndk_repository to Skylark and store it in rules_android, and decouple the rule development from Bazel core.
2) Implement support with the next revision as soon as the beta is released. I lagged behind for r17, as I only implemented support after the actual release.
3) Bundle BUILD files within the NDK. This involves a lot more effort, but could be simplified with a generation tool.

**Pull request closed without merging**